airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/cursor/file_based_cursor.py

```diff
@@ -0,0 +1,56 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Iterable, Mapping
+
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+from airbyte_cdk.sources.file_based.types import StreamState
+
+
+class FileBasedCursor(ABC):
+    """
+    Abstract base class for cursors used by file-based streams.
+    """
+
+    @abstractmethod
+    def add_file(self, file: RemoteFile) -> None:
+        """
+        Add a file to the cursor. This method is called when a file is processed by the stream.
+        :param file: The file to add
+        """
+        ...
+
+    @abstractmethod
+    def set_initial_state(self, value: StreamState) -> None:
+        """
+        Set the initial state of the cursor. The cursor cannot be initialized at construction time because the stream doesn't know its state yet.
+        :param value: The stream state
+        """
+
+    @abstractmethod
+    def get_state(self) -> Mapping[str, Any]:
+        """
+        Get the state of the cursor.
+        """
+        ...
+
+    @abstractmethod
+    def get_start_time(self) -> datetime:
```

Comment:
It feels to me that we should avoid to expose internals outside of the class. Would it make more sense to have a method `should_be_synced` like we added for the declarative package?

Reply:
The reason the start_time is exposed is that it's possible specific implementation of the StreamReader can leverage the start time to only list recent files (https://github.com/airbytehq/airbyte/pull/27382/files#diff-e07a7a23c8203f8a61e7c9d31cb495d6d9f8c7e1356ddff25630e704431d404cR127). A `should_be_synced` interface wouldn't allow that kind of optimization.

That being said, it doesn't seem like it's something we can do for S3 so I can remove it from the interface and we can revisit if we ever need the start date.